de.fhdo.lemma.servicedsl.extractor/src/de/fhdo/lemma/servicedsl/extractor/ServiceDslExtractor.xtend

```diff
@@ -0,0 +1,202 @@
+        if(importTypeKeyword === "technology") IMPORTED_TECH_ALIASES.add(^import.name)
+
+        return '''import «importTypeKeyword» from "«^import.importURI»" as «^import.name»'''
+    }
+
+    private def generate(Microservice service) {
+        val techImports = generateTechAnnotation(service)
+        val preamble = '''«service.visibility.generate» «service.type.generate»'''
+        val hasInterfaceOperations = service.interfaces.exists[!operations.empty]
+        return '''
+        «techImports»
+        «preamble» microservice «service.lemmaName» {
+            «IF hasInterfaceOperations»
+                «FOR iface : service.interfaces»
+                    «iface.generate»
+                «ENDFOR»
+            «ELSE»
+                [DEFINE_OPERATIONS]
+            «ENDIF»
+        }'''
+    }
+
+    private def lemmaName(Microservice service) {
+        return if (service.name.matches(QUALIFIED_NAME_WITH_AT_LEAST_ONE_LEVEL_PATTERN))
+                service.name
+            else
+                '''ADD_QUALIFYING_PART.«service.name»'''
+    }
+
+    private def generate(Visibility visibility) {
+        return switch(visibility) {
+            case INTERNAL: 'internal'
+            case PUBLIC: 'public'
+            default: 'architecture'
+        }
+    }
+
+    private def generate(MicroserviceType type) {
+        return switch(type) {
+            case FUNCTIONAL: 'functional'
+            case INFRASTRUCTURE: 'infrastructure'
+            case UTILITY: 'utility'
+            default: null
+        }
+    }
+
+    private def generate(Interface iface) {
+        return '''
+        interface «iface.name» {
+            «FOR o: iface.operations»
+                «o.generate»
+            «ENDFOR»
+        }'''
+    }
+
+    private def generate(Endpoint endpoint) {
+        return '''«FOR ep: endpoint.protocols SEPARATOR '; '»
+        «ep.importedProtocol.generate»:«ENDFOR
+        »«FOR ea: endpoint.addresses SEPARATOR ', '»"«ea»"«ENDFOR»;'''
+    }
+
+    private def generate(Operation operation) {
+        val endpoint =
+        '''
+        @endpoints(«FOR e: operation.endpoints»«e.generate»«ENDFOR»)
+        '''
+        val aspects = '''
+        «FOR  a: operation.aspects»«a.generate»«ENDFOR»
+        '''
+        val parameters = String.join(", ", operation.parameters.map[generate])
+
+        val comment = if(operation.apiOperationComment !== null) {
+            '''
+            ---
+            «operation.apiOperationComment.comment»
+            «FOR param : operation.parameters.filter[!it.isOptional]»
+            @required «param.name»
+            «ENDFOR»
+            ---
+            '''
+        }
+        return '''«comment» «endpoint» «aspects» «operation.name»(«parameters»);'''
+    }
+
+    private def generate(Parameter parameter) {
+        val preamble = <String>newArrayList
+        preamble.add(parameter.communicationType.generate)
```

Comment:
Introduce two variables for the generated communication type and exchange pattern, and get rid of the `String.join` statement in the template. Consequently, all parts of the template are template variables, whereby the template gets consistent and better understandable.